README.md
# fedwire
[![Build Status](https://travis-ci.org/narmitech/fedwire.svg?branch=master)](https://travis-ci.org/narmitech/fedwire)

A python package that implements an interface to write files for the [Fedwire Funds Service](https://www.frbservices.org/financial-services/wires/index.html), [a real-time gross settlement funds transfer system operated by the United States Federal Reserve Banks](https://en.wikipedia.org/wiki/Fedwire). These [compatible files](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.445.7645&rep=rep1&type=pdf) include routing instructions that, once received and processed, will debit the funds from the sending bank's reserve account at their Federal Reserve bank and credit the receiving bank's account. Wire transfers sent via Fedwire are completed in the same day, while some are completed instantly.

## Usage

```
from datetime import datetime
from decimal import Decimal

from fedwire import FedwireFile, Entry, Tag

wire_file = FedwireFile()
entries = [
    Entry([
      Tag.sender_supplied_information(production=True),
      Tag.type(Tag.TYPE_FUNDS_TRANSFER, TAG.SUBTYPE_BASIC_FUNDS_TRANSFER),
      Tag.imad(datetime.now(), source, sequence),
      Tag.amount(Decimal('123')),
      Tag.sender_institution(routing, name),
      Tag.receiver_institution(routing, name),
      Tag.business_function_code(business, transaction),
    ])
]
wire_file.add_batch(entries)
print(wire_file)
```

## Format details

The Fedwire Funds Service message format consists of tags and data elements within those tags. Elements are either fixed or variable in length, with a maximum number of characters allotted for each element.

## License

Apache License 2.0 See LICENSE for details.


CHANGELOG.md
# Change Log

## [v0.1.2](https://github.com/narmitech/fedwire-python/tree/v0.1.2)
[Full Changelog](https://github.com/narmitech/fedwire-python/compare/v0.1.1...v0.1.2)

* Fix amount and make tests more explicit (https://github.com/narmitech/fedwire/pull/6)

## [v0.1.1](https://github.com/narmitech/fedwire-python/tree/v0.1.1)
[Full Changelog](https://github.com/narmitech/fedwire-python/compare/v0.1.0...v0.1.1)

* Fix originator tag (https://github.com/narmitech/fedwire/pull/2)

## [v0.1.0](https://github.com/narmitech/fedwire-python/tree/v0.1.0)
[Full Changelog](https://github.com/narmitech/fedwire-python/compare/...v0.1.0)

* Add interfaces


tests/test_format.py
from datetime import datetime
from decimal import Decimal
from unittest import TestCase

from fedwire import Entry, Tag


class TestEntry(TestCase):
    def setUp(self):
        self.amount = '10.00'
        self.sender = '211111086'
        self.receiver = '011111390'
        self.entry = Entry([
            Tag.sender_supplied_information(production=True),
            Tag.type(Tag.TYPE_FUNDS_TRANSFER, Tag.SUBTYPE_BASIC_FUNDS_TRANSFER),
            Tag.imad(datetime.now(), '4', ''),
            Tag.amount(Decimal(self.amount)),
            Tag.sender_institution(self.sender, ''),
            Tag.receiver_institution(self.receiver, 'Bank of America*'),
            Tag.business_function_code(Tag.CODE_CTR),
        ])

    def test_line(self):
        line = self.entry.line
        expected_line = '{1500}30        P {1510}1000{1520}20191210       4      {2000}000000001000{3100}211111086*{3400}011111390Bank of America**{3600}CTR'  # noqa
        self.assertEqual(expected_line, line)

    def test_invalid(self):
        empty = Entry([])
        self.assertFalse(empty.is_valid())

    def test_invalid_tag(self):
        self.entry.tags.pop()
        self.entry.tags.append(Tag.business_function_code('asdffoobarbaz'))
        self.assertFalse(self.entry.is_valid())

    def test_valid(self):
        self.assertTrue(self.entry.is_valid())


class TestTag(TestCase):
    def test_str(self):
        result = Tag('0000', 'value', 10)
        self.assertEqual('{0000}value', str(result))

    def test_sender_supplied_information(self):
        tag = Tag.sender_supplied_information(production=True)
        self.assertEqual('{1500}30        P ', str(tag))

    def test_type(self):
        tag = Tag.type(Tag.TYPE_FUNDS_TRANSFER, Tag.SUBTYPE_BASIC_FUNDS_TRANSFER)
        self.assertEqual('{1510}1000', str(tag))

    def test_imad(self):
        now = datetime.now()
        tag = Tag.imad(now, '4', '')
        self.assertEqual('{}{}       4      '.format('{1520}', now.strftime('%Y%m%d')), str(tag))

    def test_amount(self):
        tag = Tag.amount(Decimal('10.00'))
        self.assertEqual('{2000}000000001000', str(tag))

    def test_sender_institution(self):
        tag = Tag.sender_institution('211111086', '')
        self.assertEqual('{3100}211111086*', str(tag))

    def test_receiver_institution(self):
        tag = Tag.receiver_institution('011111390', 'Bank of America')
        self.assertEqual('{3400}011111390Bank of America*', str(tag))

    def test_business_function_code(self):
        tag = Tag.business_function_code(Tag.CODE_CTR)
        self.assertEqual('{3600}CTR', str(tag))

    def test_sender_reference(self):
        tag = Tag.sender_reference('111113')
        self.assertEqual('{3320}111113*', str(tag))

    def test_beneficiary(self):
        tag = Tag.beneficiary(
            Tag.ID_DEMAND_DEPOSIT_ACCOUNT_NUMBER,
            '987654111',
            'Test Guy 1',
            '1 GONE Street\n\nCharlestown, MA 02129')
        expected = '{4200}D987654111*Test Guy 1*1 GONE Street*Charlestown, MA 02129*'
        self.assertEqual(expected, str(tag))

    def test_originator(self):
        tag = Tag.originator(
            Tag.ID_DEMAND_DEPOSIT_ACCOUNT_NUMBER,
            '6111111113',
            'KEVIN DOUGH',
            '8 FOOD RD\nSOMEWHERE, MA 10821')
        self.assertEqual('{5000}D6111111113*KEVIN DOUGH*8 FOOD RD*SOMEWHERE, MA 10821*', str(tag))

    def test_originator_to_beneficiary(self):
        tag = Tag.originator_to_beneficiary('Test 2')
        self.assertEqual('{6000}Test 2*', str(tag))


tests/test_generator.py
from datetime import datetime
from decimal import Decimal
from unittest import TestCase

from fedwire import Entry, Tag, FedwireFile


class TestFedwireFile(TestCase):
    def setUp(self):
        self.entries = [
            Entry([
                Tag.sender_supplied_information(production=True),
                Tag.type(Tag.TYPE_FUNDS_TRANSFER, Tag.SUBTYPE_BASIC_FUNDS_TRANSFER),
                Tag.imad(datetime.now(), '4', ''),
                Tag.amount(Decimal('10.00')),
                Tag.sender_institution('211111086', ''),
                Tag.sender_reference('111113'),
                Tag.receiver_institution('011111390', 'Bank of America'),
                Tag.business_function_code(Tag.CODE_CTR),
                Tag.beneficiary(
                  Tag.ID_DEMAND_DEPOSIT_ACCOUNT_NUMBER,
                  '987654111',
                  'Test Guy 1',
                  '1 Walker Street\n\nCharlestown, MA 02129'),
                Tag.originator(
                    Tag.ID_DEMAND_DEPOSIT_ACCOUNT_NUMBER,
                    '6111111113',
                    'KEVIN DOUGH',
                    '8 FOOD RD\n\nSOMEWHERE, MA 10821'),
                Tag.originator_to_beneficiary('Test 2'),
            ])
        ]

    def test_add_batch(self):
        wire_file = FedwireFile()
        wire_file.add_batch(self.entries)
        wire_file.add_batch(self.entries)
        self.assertEqual(len(wire_file.entries), 2)

    def test_str(self):
        wire_file = FedwireFile(entries=self.entries)
        self.assertEqual(str(wire_file).count('\n'), 1)


fedwire/format.py
from decimal import Decimal, ROUND_HALF_UP


class BaseFedwireException(Exception):
    """
    Base class for all fedwire exceptions
    """
    pass


class FedwireConfigurationException(BaseFedwireException):
    """
    Class for API Configuration Exceptions
    """
    pass


TAG_SENDER_SUPPLIED_INFORMATION = 1500
TAG_TYPE = 1510
TAG_IMAD = 1520  # Input Message Accountability Data
TAG_AMOUNT = 2000
TAG_SENDER_DEPOSITORY_INSTITUTION = 3100
TAG_SENDER_REFERENCE = 3320
TAG_RECEIVER_DEPOSITORY_INSTITUTION = 3400
TAG_BUSINESS_FUNCTION_CODE = 3600
TAG_BENEFICIARY = 4200
TAG_ORIGINATOR = 5000
TAG_ORIGINATOR_TO_BENEFICIARY = 6000

FORMAT_VERSION = '30'
ENVIRONMENT_TEST = 'T'
ENVIRONMENT_PRODUCTION = 'P'
DUPLICATION_ORIGINAL = ' '
DUPLICATION_RESEND = 'P'

# See https://fraser.stlouisfed.org/files/docs/historical/ny%20circulars/nycirc_1995_10759.pdf
# page 16 of PDF, labeled page 14
VARIABLE_LENGTH_DELIMITER = '*'


class Entry:
    REQUIRED_TAGS = [
        TAG_SENDER_SUPPLIED_INFORMATION,
        TAG_TYPE,
        TAG_IMAD,
        TAG_AMOUNT,
        TAG_SENDER_DEPOSITORY_INSTITUTION,
        TAG_RECEIVER_DEPOSITORY_INSTITUTION,
        TAG_BUSINESS_FUNCTION_CODE,
    ]

    def __init__(self, tags=[]):
        self.tags = tags

    @property
    def line(self):
        sorted_tags = sorted(self.tags, key=lambda t: t.name)
        return ''.join([str(t) for t in sorted_tags])

    def is_valid(self):
        for tag in self.tags:
            if not tag.is_valid():
                return False

        tag_names = [tag.name for tag in self.tags]
        if len(set(self.REQUIRED_TAGS) - set(tag_names)) > 0:
            return False

        return True


class Tag:
    CODE_BTR = 'BTR'  # Bank Transfer (Beneficiary is a bank)
    CODE_CKS = 'CKS'  # Check Same Day Settlement
    CODE_CTP = 'CTP'  # Customer Transfer Plus
    CODE_CTR = 'CTR'  # Customer Transfer (not to a bank) Deposit to Sender's Account
    CODE_DRC = 'DRC'  # Customer or Corporate Drawdown Request
    CODE_FFR = 'FFR'  # Fed Funds Returned
    CODE_FFS = 'FFS'  # Fed Funds Sold

    """
    A funds transfer in which the sender and/or receiver may be a bank or a
    third party (i.e., customer of a bank).
    """
    TYPE_FUNDS_TRANSFER = '10'
    """
    A funds transfer to or from a foreign central bank or government or
    international organization with an account at the Federal Reserve Bank of
    New York.
    """
    TYPE_FOREIGN_TRANSFER = '15'
    """
    A funds transfer between Fedwire Funds Service participants.
    """
    TYPE_SETTLEMENT_TRANSFER = '16'
    """
    A basic value funds transfer.
    """
    SUBTYPE_BASIC_FUNDS_TRANSFER = '00'
    """
    A non-value request for reversal of a funds transfer originated on the
    current business day.
    """
    SUBTYPE_REQUEST_FOR_REVERSAL = '01'
    """
    A value reversal of a funds transfer received on the current business day.
    May be used in response to a subtype code ‘01’ Request for Reversal.
    """
    SUBTYPE_REVERSAL_OF_TRANSFER = '02'
    """
    A non-value request for a reversal of a funds transfer originated on a
    prior business day.
    """
    SUBTYPE_REQUEST_FOR_REVERSAL_PRIOR_DAY_TRANSFER = '07'
    """
    A value reversal of a funds transfer received on a prior business day. May
    be used in response to a subtype code ‘07’ Request for Reversal of a Prior
    Day Transfer.
    """
    SUBTYPE_REVERSAL_PRIOR_DAY_TRANSFER = '08'
    """
    A non-value request for the receiver to send a funds transfer to a
    designated party.
    """
    SUBTYPE_REQUEST_CREDIT = '31'
    """
    A value funds transfer honoring a subtype 31 request for credit.
    """
    SUBTYPE_FUNDS_TRANSFER_HONOR_REQUEST_FOR_CREDIT = '32'
    """
    A non-value message indicating refusal to honor a subtype 31 request for
    credit.
    """
    SUBTYPE_REFUSAL_HONOR_REQUEST_FOR_CREDIT = '33'
    """
    A non-value message used to communicate questions and information that is
    not covered by a specific subtype.
    """
    SUBTYPE_SERVICE_MESSAGE = '90'

    ID_SWIFT_BANK_IDENTIFIER_CODE = 'B'  # BIC
    ID_CHIPS_PARTICIPANT = 'C'
    ID_DEMAND_DEPOSIT_ACCOUNT_NUMBER = 'D'  # DDA
    ID_FED_ROUTING_NUMBER = 'F'
    ID_SWIFT_BIC_OR_BANK_ENTITY_IDENTIFIER_AND_ACCOUNT_NUMBER = 'T'  # BEI
    ID_CHIPS_IDENTIFIER = 'U'
    ID_PASSPORT_NUMBER = '1'
    ID_TAX_IDENTIFICATION_NUMBER = '2'
    ID_DRIVERS_LICENSE_NUMBER = '3'
    ID_ALIEN_REGISTRATION_NUMBER = '4'
    ID_CORPORATE_IDENTIFICATION = '5'
    ID_OTHER_IDENTIFICATION = '9'

    @classmethod
    def sender_supplied_information(klass, production=False, resend=False):
        user_request_correlation = make_space(8)  # all spaces is blank
        environment = ENVIRONMENT_PRODUCTION if production else ENVIRONMENT_TEST
        duplication = DUPLICATION_RESEND if resend else DUPLICATION_ORIGINAL
        return klass(
            TAG_SENDER_SUPPLIED_INFORMATION,
            FORMAT_VERSION + user_request_correlation + environment + duplication,
            12)

    @classmethod
    def type(klass, transfer_type, transfer_subtype):
        return klass(TAG_TYPE, transfer_type + transfer_subtype, 4)

    @classmethod
    def imad(klass, input_date, source, sequence):
        # YYYYMMDD
        # Input Source (8 characters)
        # Input Sequence Number (6 characters)
        value = input_date.strftime('%Y%m%d') + \
            (make_space(8 - len(source)) + source) + \
            (make_space(6 - len(sequence)) + sequence)
        return klass(TAG_IMAD, value, 22)

    @classmethod
    def amount(klass, value):
        # 12 numeric, right-justified with leading zeros, an implied decimal
        # point and no commas; e.g., $12,345.67 becomes 000001234567
        cents = Decimal('0.01')
        amount = value.quantize(cents, ROUND_HALF_UP)
        formatted_amount = str(amount).replace('.', '').zfill(12)
        return klass(TAG_AMOUNT, formatted_amount, 12)

    @classmethod
    def sender_institution(klass, routing, name):
        # Sender ABA Number (9 characters)
        # Sender Short Name (18 characters)
        return klass(TAG_SENDER_DEPOSITORY_INSTITUTION, mark_variable(routing + name, 27), 27)

    @classmethod
    def receiver_institution(klass, routing, name):
        return klass(TAG_RECEIVER_DEPOSITORY_INSTITUTION, mark_variable(routing + name, 27), 27)

    @classmethod
    def business_function_code(klass, business, transaction=''):
        return klass(TAG_BUSINESS_FUNCTION_CODE, business + transaction, 6)

    @classmethod
    def sender_reference(klass, value):
        return klass(TAG_SENDER_REFERENCE, mark_variable(value, 16), 16)

    @classmethod
    def beneficiary(klass, code, identifier, name, address):
        # ID Code (B, C, D, F, T, U, 1, 2, 3, 4, 5, 9)
        # Identifier (34 characters)
        # Name (35 characters)
        # Address (3 lines of 35 characters each)
        marked_identifier = mark_variable(identifier, 34)
        marked_name = mark_variable(name, 35)
        address_lines = address.split('\n')
        marked_address = ''.join([mark_variable(l, 35) for l in address_lines])
        return klass(TAG_BENEFICIARY, code + marked_identifier + marked_name + marked_address, 175)

    @classmethod
    def originator(klass, code, identifier, name, address):
        # ID Code (B, C, D, F, T, U, 1, 2, 3, 4, 5, 9)
        # Identifier (34 characters)
        # Name (35 characters)
        # Address (3 lines of 35 characters each)
        marked_identifier = mark_variable(identifier, 34)
        marked_name = mark_variable(name, 35)
        address_lines = address.split('\n')
        marked_address = ''.join([mark_variable(l, 35) for l in address_lines])
        return klass(TAG_ORIGINATOR, code + marked_identifier + marked_name + marked_address, 175)

    @classmethod
    def originator_to_beneficiary(klass, value):
        # up to 4 lines of 35 characters each
        lines = value.split('\n')
        marked_lines = [mark_variable(l, 35) for l in lines]
        return klass(TAG_ORIGINATOR_TO_BENEFICIARY, ''.join(marked_lines), 140)

    def __init__(self, name, value, max_length):
        self.name = name
        self.value = str(value)
        self.max_length = max_length

    def is_valid(self):
        return len(self.value) <= self.max_length

    def __str__(self):
        value = ''.join(self.value)
        # Double the curly braces to escape them. :shrug:
        return '{{{}}}{}'.format(self.name, value)


def make_space(space_padding=0):
    """
    Return string with x number of spaces. Defaults to 0.
    """
    space = ''

    for i in range(space_padding):
        space += ' '

    return space


def mark_variable(string, max_length=0):
    if len(string) == 0 or len(string) > max_length:
        return string

    return string + VARIABLE_LENGTH_DELIMITER
